app/routes/docs.$lang.$ref.$.tsx

```diff
@@ -59,6 +59,28 @@ export const headers: HeadersFunction = ({ loaderHeaders }) => {
 
 const LAYOUT_LOADER_KEY = "routes/docs.$lang.$ref";
 
+function validateParentData(data: unknown): data is {
+  latestVersion: string;
+  releaseBranch: string;
+  branches: string[];
+  currentGitHubRef: string;
+} {
+  return (
+    !!data &&
+    typeof data === "object" &&
+    "latestVersion" in data &&
+    "releaseBranch" in data &&
+    "branches" in data &&
+    "currentGitHubRef" in data
+  );
+}
+
+function hasIsProductionHost(
+  data: unknown,
+): data is { isProductionHost: boolean } {
+  return !!data && typeof data === "object" && "isProductionHost" in data;
+}
+
 export const meta: MetaFunction<
```

Comment:
Alright, I just realized something. We're actually declaring the types of `matches` data and passing it into `MetaFunction`, however those types aren't preserved by `getMatchesData`. If we just update this slightly we get all of the types we want. Since they're coming from the actual types of the `loaders`, this is about as much type safety as we need

```ts
const LAYOUT_LOADER_KEY = "routes/docs.$lang.$ref";

type Loader = typeof loader;
type MatchLoaders = {
  [LAYOUT_LOADER_KEY]: typeof docsLayoutLoader;
  root: typeof rootLoader;
};
export const meta: MetaFunction<Loader, MatchLoaders> = (args) => {
  let { data } = args;

  let matchesData = getMatchesData<Loader, MatchLoaders>(args);
```

`matchesData` will be typed properly at this point.

I think this is the only suggestion I have at this point, otherwise I'm good to merge it in!

Reply:
Thank you for your suggestion.
Passing types to `getMatchesData` makes the code simple, it is very nice!